hbase-server/src/main/java/org/apache/hadoop/hbase/master/AssignmentManager.java

```diff
@@ -2079,16 +2085,29 @@ private void unassign(final HRegionInfo region,
         }
 
         if (logRetries) {
-          LOG.info("Server " + server + " returned " + t + " for "
-            + region.getRegionNameAsString() + ", try=" + i
-            + " of " + this.maximumAttempts, t);
+          LOG.info("Server " + server + " returned " + t + " for " + region.getRegionNameAsString()
+              + ", try=" + i + " of " + this.maximumAttempts,
+            t);
           // Presume retry or server will expire.
         }
       }
     }
-    // Run out of attempts
-    if (state != null) {
-      regionStates.updateRegionState(region, State.FAILED_CLOSE);
+
+    long sleepTime = backoffPolicy.getBackoffTime(retryConfig,
```

Comment:
Pardon my ignorance but I don't fully understand the fix. I think the ask in the jira to spread out the unassigns a bit by using a backoff based approach. To do that why not just fix the sleepTime above in L2075 to use an exponential backoff based approach?

Reply:
The idea is to use the exponential backoff configs "hbase.assignment.retry.sleep.initial" and "hbase.assignment.retry.sleep.initial" for backoff between retries as they can be exhausted pretty fast in case where the server is loaded /busy and cannot really even acknowledge the region close request from the master.  We need to use them to schedule the retry at a later point in a different thread asynchronouly

The sleepTime is not really meant for this use case and is not reading any exponential backoff configs

Reply:
Right, so my point is this. If there is an exception in sendRegionClose() (say ServerTooBusy), we enter the catch() block in L2024, we don't match the if / else if in L2031 (failed server) / 2038 (region not serving) and that leaves us with just looping for "maximumAttempts" quickly in succession without a backoff (this is the problem we are trying to address). My point was why not do something like this.

```
 if (t instanceof RegionServerAbortedException || t instanceof RegionServerStoppedException
            || t instanceof RegionServerStoppedException	
            || t instanceof ServerNotRunningYetException) {	        
......
} else if ((t instanceof FailedServerException)....) {
.....
} else {
 // Handle any other remote exception
 // use your exponential backoff algorithm...
} 
```

Basically we have the loop for maxAttempts just that it is not using an exponential backoff. Just plug it in an else {} block and we are good? What is advantage of doing it with a delayed callable, doesn't that just complicate the state machine with async requests and book-keeping or is there something that I'm missing?

Reply:
There are two reasons for this:
1. It also deals with the case where the sendRegionClose() itself returns false without getting into the catch block though I am not very sure of the case where it would reach there. the method doc says the method returns false if it doesn't acknowledge the request
2. The current sleepTime in the else of catch block is static at 100ms. The exponential backoff is configurable and can even be 5 minutes , so there is no point holding the thread for that amount of time  and rather have it deal asynchronously.

I cannot think of any state machine issue as even if just sleep the current thread , the state would remain the same and the method as such is not returning anything and anyone else would just check the state change made by the method either synchronously or asynchronously. Rather we would delegate it to a new thread and not block the current thread.

Reply:
The current code does the sleep in the current thread where it is handling the region already in transition kind of case where there is a max wait time and hence they just sleep for every 100ms and keep retrying before the max ttime ends. In our case we definitely know that we don't want to retry before a configured backoff time, so we can as well schedule it later